esc_controlled.rs

```diff
@@ -0,0 +1,386 @@
+
+        let phase1 = Pwm::new(cx.device.PWM0);
+        phase1
+            .set_output_pin(pwm::Channel::C0, p1.high_side)
+            .set_output_pin(pwm::Channel::C1, p1.low_side)
+            .set_period(1u32.mhz().into())
+            .set_duty(0.1);
+
+        let phase2 = Pwm::new(cx.device.PWM1);
+        phase2
+            .set_output_pin(pwm::Channel::C0, p2.high_side)
+            .set_output_pin(pwm::Channel::C1, p2.low_side)
+            .set_period(1u32.mhz().into())
+            .set_duty(0.1);
+        let phase3 = Pwm::new(cx.device.PWM2);
+        phase3
+            .set_output_pin(pwm::Channel::C0, p3.high_side)
+            .set_output_pin(pwm::Channel::C1, p3.low_side)
+            .set_period(1u32.mhz().into())
+            .set_duty(0.1);
+
+        //  The order that we should drive the phases.
+        let drive_pattern = DrivePattern::new();
+
+        let sender = protocol::sender::Sender::new();
+        debug_assert!(Mono::now().duration_since_epoch().to_micros() != 0);
+
+        let (current_sense_sender, current_sense_rec) = rtic_sync::make_channel!((u64, f32), 10);
+        // Spawn the tasks.
+        motor_driver::spawn().ok().unwrap();
+        control_loop::spawn().ok().unwrap();
+        (
+            Shared {
+                // Initialization of shared resources go here
+                drive_pattern,
+                sender,
+                duty: 0.2,
+                phase1,
+                phase2,
+                phase3,
+                velocity: 0.,
+                current_target: 0.2,
+            },
+            Local {
+                // Initialization of local resources go here
+                events,
+                current_sense,
+                current_sense_sender,
+                current_sense_rec,
+                hal_pins,
+            },
+        )
+    }
+
+    #[task(binds=GPIOTE, local=[events,hal_pins,t:Option<u64> = None],shared = [drive_pattern,phase1,phase2,phase3,duty,velocity], priority = 5)]
+    /// Manages gpiote interrupts.
+    ///
+    /// This is every pin interrupt. So hal effect feedback, can bus messages
+    /// and possibly more. the contents of this function will vary with the
+    /// board we are using.
+    ///
+    /// ## Motor driving
+    ///
+    /// This function also drives the motors. This is the defualt case. The only
+    /// other time that the motor should be driven externally is if the PID
+    /// controller requests a new velocity.
+    ///
+    /// This ensures that the motor can be started from standstill.
+    fn handle_gpio(mut cx: handle_gpio::Context) {
+        defmt::info!("GPIO interrupt");
+        let t = Mono::now();
+        // Docs says that it is infallible.
+        let edge =
+            |pin: &mut Pin<Input<PullUp>>| -> bool { unsafe { pin.is_high().unwrap_unchecked() } };
+        let mut set_phase = false;
+        let mut positive_flank = false;
+        for event in cx.local.events.events() {
```

Comment:
Makes sense to iterate over the `gpiote channels`. However, would this increase the overhead if we do it sequentially together with the motor driving and PWM? Is it possible to separate into multiple GPIO tasks (just a general thought).  Looks really good in general!